File: rust/src/types/a2a.rs
//! A2A verification context types (v1.4).
//!
//! SchemaPin verifies tool schemas. When a tool invocation crosses an A2A
//! (Agent-to-Agent) trust boundary, the verifier needs to scope the result
//! to the intersection of *caller-trusted* domains and the *tool provider's*
//! domain. This module defines [`A2aVerificationContext`] and the
//! [`AllowedDomains`] helpers used to compute that intersection.
//!
//! ## AllowedDomains convention
//!
//! The allow-list semantics mirror AgentPin v0.3 (§4.11 of the AgentPin
//! technical specification) **exactly**:
//!
//! > **An empty `AllowedDomains` list means *unrestricted* (all domains
//! > trusted).**
//!
//! This is the opposite of the naïve set-theoretic interpretation (where an
//! empty set allows nothing) but matches the existing v1.3 behaviour where
//! an *omitted* `allowed_domains` field allowed all domains. It lets
//! callers pass `Vec::new()` to mean "no restriction" without inventing a
//! sentinel.
//!
//! SchemaPin defines these helpers locally rather than depending on the
//! AgentPin crate — keeping SchemaPin self-contained for users who only
//! consume tool-integrity, and avoiding a circular trust-stack dependency.
//! Callers who *do* link both SDKs can pass the result of
//! `agentpin::AllowedDomains::intersect(...)` into
//! [`A2aVerificationContext::trusted_domains`] verbatim — the wire and
//! in-memory shapes are identical.

use serde::{Deserialize, Serialize};

/// Information that scopes a schema verification to an A2A interaction.
///
/// Pair with [`crate::verification::verify_schema_for_a2a`] to run the
/// standard 7-step verification flow with the additional A2A scope check.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct A2aVerificationContext {
    /// Caller's agent identity (URN-style, matching AgentPin).
    ///
    /// Informational only — SchemaPin does not validate the URN shape or
    /// check it against an external identity provider. The field exists so
    /// downstream policy engines (Symbiont) can correlate verification
    /// results with the originating caller.
    pub caller_agent_id: String,

    /// Depth in the A2A delegation chain. `0` means direct caller.
    ///
    /// Verifiers SHOULD reject `delegation_depth > 3` to match the AgentPin
    /// `max_delegation_depth` cap (AgentPin spec §4.3). The check is
    /// performed inside [`crate::verification::verify_schema_for_a2a`].
    pub delegation_depth: u8,

    /// Originating domain of the A2A request. Informational; the scope check
    /// uses `trusted_domains` ∩ {tool provider domain}, not this field.
    pub originating_domain: String,

    /// Caller-trusted domains.
    ///
    /// Uses the AgentPin convention: **an empty `Vec` means *unrestricted***
    /// (all domains trusted), not "deny-all". See module-level docs.
    pub trusted_domains: Vec<String>,
}

impl A2aVerificationContext {
    /// Construct a context that places no restriction on which provider
    /// domains may be verified through.
    pub fn unrestricted(caller_agent_id: impl Into<String>) -> Self {
        Self {
            caller_agent_id: caller_agent_id.into(),
            delegation_depth: 0,
            originating_domain: String::new(),
            trusted_domains: Vec::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// AllowedDomains helpers
// ---------------------------------------------------------------------------
//
// Spec source of truth: AgentPin technical specification §4.11. This module
// re-implements the helpers locally rather than linking the AgentPin crate
// (see module-level docs for rationale). Update both projects in lockstep if
// the convention ever changes.

/// Returns `true` when `list` is an unrestricted allow-list (empty).
pub fn is_unrestricted(list: &[String]) -> bool {
    list.is_empty()
}

/// Returns `true` when `domain` is permitted under `list`.
///
/// An empty `list` allows everything. A non-empty `list` allows the domain
/// when it exactly matches an entry OR matches one of the entry's wildcard
/// patterns. Pattern matching follows AgentPin spec §5.5: a leading `*.`
/// matches any subdomain (e.g. `*.client.com` matches `api.client.com`
/// but not `client.com` itself).
pub fn allows(list: &[String], domain: &str) -> bool {
    if is_unrestricted(list) {
        return true;
    }
    list.iter().any(|pattern| pattern_matches(pattern, domain))
}

/// Intersection of two allow-lists, honouring the `empty = unrestricted`
/// convention.
///
/// - `unrestricted ∩ X = X`
/// - `X ∩ unrestricted = X`
/// - Otherwise: literal set intersection of `lhs` and `rhs` (string equality;
///   no wildcard expansion). Preserves the order of `lhs`.
///
/// An intersection that yields an empty `Vec` from two non-empty inputs is
/// *re-interpreted as unrestricted* under the same convention. This is the
/// AgentPin spec §4.11.4 edge case — callers needing to distinguish
/// "intentionally restricted to nothing" from "no restriction" must track
/// that distinction outside the allow-list value.
pub fn intersect(lhs: &[String], rhs: &[String]) -> Vec<String> {
    if is_unrestricted(lhs) {
        return rhs.to_vec();
    }
    if is_unrestricted(rhs) {
        return lhs.to_vec();
    }
    lhs.iter()
        .filter(|d| rhs.iter().any(|r| r == *d))
        .cloned()
        .collect()
}

/// Wildcard-aware domain pattern matcher.
///
/// `*.example.com` matches `api.example.com` and `a.b.example.com` but NOT
/// `example.com` itself. Anything without a leading `*.` is a literal
/// equality check.
fn pattern_matches(pattern: &str, domain: &str) -> bool {
    if let Some(suffix) = pattern.strip_prefix("*.") {
        // Match {anything}.suffix but not bare suffix.
        if domain.len() <= suffix.len() {
            return false;
        }
        let dot_position = domain.len() - suffix.len() - 1;
        return domain.ends_with(suffix) && domain.as_bytes().get(dot_position) == Some(&b'.');
    }
    pattern == domain
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unrestricted_helpers_round_trip() {
        let ad: Vec<String> = Vec::new();
        assert!(is_unrestricted(&ad));
        assert!(allows(&ad, "anything.example.com"));
        assert!(allows(&ad, "literally-anything"));
    }

    #[test]
    fn restricted_allows_only_listed_domains() {
        let ad: Vec<String> = vec!["api.client.com".to_string(), "*.partner.com".to_string()];
        assert!(allows(&ad, "api.client.com"));
        assert!(allows(&ad, "tools.partner.com"));
        assert!(allows(&ad, "deep.nested.partner.com"));
        assert!(!allows(&ad, "partner.com")); // *.partner.com does not match bare partner.com
        assert!(!allows(&ad, "evil.example.com"));
    }

    #[test]
    fn intersect_with_unrestricted_returns_other() {
        let unrestricted: Vec<String> = Vec::new();
        let restricted = vec!["a.com".to_string(), "b.com".to_string()];
        assert_eq!(intersect(&unrestricted, &restricted), restricted);
        assert_eq!(intersect(&restricted, &unrestricted), restricted);
    }

    #[test]
    fn intersect_returns_overlap() {
        let lhs = vec![
            "a.com".to_string(),
            "b.com".to_string(),
            "c.com".to_string(),
        ];
        let rhs = vec![
            "b.com".to_string(),
            "c.com".to_string(),
            "d.com".to_string(),
        ];
        assert_eq!(
            intersect(&lhs, &rhs),
            vec!["b.com".to_string(), "c.com".to_string()]
        );
    }

    #[test]
    fn intersect_empty_overlap_is_treated_as_unrestricted() {
        let lhs = vec!["a.com".to_string()];
        let rhs = vec!["b.com".to_string()];
        let result = intersect(&lhs, &rhs);
        assert!(is_unrestricted(&result));
    }

    #[test]
    fn pattern_wildcard_does_not_match_bare_domain() {
        assert!(pattern_matches("*.partner.com", "api.partner.com"));
        assert!(!pattern_matches("*.partner.com", "partner.com"));
        assert!(!pattern_matches("*.partner.com", "partnercom"));
    }
}

File: go/pkg/verification/a2a.go
package verification

// A2A verification context for SchemaPin (v1.4 alpha.3).
//
// Mirrors the Rust schemapin::types::a2a module and the AgentPin v0.3
// AllowedDomains semantics (AgentPin technical specification §4.11).
//
// When a SchemaPin verification crosses an A2A (Agent-to-Agent) trust
// boundary, the verifier needs to scope the result to the intersection of
// caller-trusted domains and the tool provider's domain. A2AVerificationContext
// carries that scoping data; pair it with VerifySchemaForA2A.
//
// AllowedDomains convention
//
// The allow-list semantics follow AgentPin v0.3 exactly:
//
//	An empty TrustedDomains list means UNRESTRICTED — all domains trusted.
//	This is the opposite of the naïve set-theoretic interpretation but
//	matches v1.3 behaviour where an *omitted* allowed_domains field allowed
//	all domains.
//
// SchemaPin re-implements these helpers locally rather than depending on
// the agentpin Go module — keeping SchemaPin self-contained and avoiding a
// circular trust-stack dependency. Callers who *do* depend on both modules
// can pass the result of agentpin's intersection helper directly into
// TrustedDomains — the wire and in-memory shapes are identical.

// A2AMaxDelegationDepth is the maximum A2A delegation depth allowed by
// this verifier (v1.4 alpha.3).
//
// Mirrors the AgentPin max_delegation_depth cap (AgentPin spec §4.3).
// Bumping this on the SchemaPin side without a matching bump on AgentPin
// would let SchemaPin accept chains AgentPin would reject — keep in lockstep.
const A2AMaxDelegationDepth uint8 = 3

// A2AVerificationContext scopes a schema verification to an A2A interaction.
//
// Pair with VerifySchemaForA2A to run the standard 7-step verification flow
// with the additional A2A scope check.
type A2AVerificationContext struct {
	// CallerAgentID is the caller's agent identity (URN-style, matching
	// AgentPin). Informational only — SchemaPin does not validate the URN
	// shape.
	CallerAgentID string

	// DelegationDepth is the depth in the A2A delegation chain. 0 = direct
	// caller. Verifiers SHOULD reject DelegationDepth > A2AMaxDelegationDepth.
	DelegationDepth uint8

	// OriginatingDomain is the originating domain of the A2A request.
	// Informational; the scope check uses TrustedDomains vs. provider domain.
	OriginatingDomain string

	// TrustedDomains is the caller-trusted domain allow-list. Uses the
	// AgentPin convention: an empty slice means UNRESTRICTED (all domains
	// trusted), not "deny-all".
	TrustedDomains []string
}

// NewUnrestrictedA2AContext constructs a context that places no restriction
// on which provider domains may be verified through.
func NewUnrestrictedA2AContext(callerAgentID string) *A2AVerificationContext {
	return &A2AVerificationContext{
		CallerAgentID:   callerAgentID,
		DelegationDepth: 0,
	}
}

// AllowedDomains helpers
//
// Spec source of truth: AgentPin technical specification §4.11. This file
// re-implements the helpers locally rather than linking the agentpin Go
// module. Update both projects in lockstep if the convention ever changes.

// A2AIsUnrestricted returns true when list is empty (no restriction).
func A2AIsUnrestricted(list []string) bool {
	return len(list) == 0
}

// A2AAllows returns true when domain is permitted under list.
//
// An empty list allows everything. A non-empty list allows domain when it
// exactly matches an entry OR matches one of the entry's wildcard
// patterns. Pattern matching follows AgentPin spec §5.5: a leading "*."
// matches any subdomain (e.g. "*.client.com" matches "api.client.com" but
// NOT "client.com" itself).
func A2AAllows(list []string, domain string) bool {
	if A2AIsUnrestricted(list) {
		return true
	}
	for _, p := range list {
		if patternMatches(p, domain) {
			return true
		}
	}
	return false
}

// A2AIntersect returns the intersection of two allow-lists, honouring the
// empty-is-unrestricted convention.
//
//   - unrestricted ∩ X = X
//   - X ∩ unrestricted = X
//   - Otherwise: literal set intersection (string equality; no wildcard
//     expansion), preserving the order of lhs.
//
// An intersection that yields an empty slice from two non-empty inputs is
// *re-interpreted as unrestricted* under the same convention — see AgentPin
// spec §4.11.4. Callers needing to distinguish "intentionally restricted to
// nothing" from "no restriction" must track that outside the slice.
func A2AIntersect(lhs, rhs []string) []string {
	if A2AIsUnrestricted(lhs) {
		out := make([]string, len(rhs))
		copy(out, rhs)
		return out
	}
	if A2AIsUnrestricted(rhs) {
		out := make([]string, len(lhs))
		copy(out, lhs)
		return out
	}
	rhsSet := make(map[string]struct{}, len(rhs))
	for _, d := range rhs {
		rhsSet[d] = struct{}{}
	}
	out := make([]string, 0)
	for _, d := range lhs {
		if _, ok := rhsSet[d]; ok {
			out = append(out, d)
		}
	}
	return out
}

// patternMatches is a wildcard-aware domain pattern matcher.
//
// "*.example.com" matches "api.example.com" and "a.b.example.com" but NOT
// "example.com" itself. Anything without a leading "*." is a literal
// equality check.
func patternMatches(pattern, domain string) bool {
	if len(pattern) > 2 && pattern[0] == '*' && pattern[1] == '.' {
		suffix := pattern[2:]
		if len(domain) <= len(suffix) {
			return false
		}
		if domain[len(domain)-len(suffix):] != suffix {
			return false
		}
		return domain[len(domain)-len(suffix)-1] == '.'
	}
	return pattern == domain
}
